lib/internal/iterator.js

```diff
@@ -0,0 +1,38 @@
+'use strict';
+
+import isArrayLike from 'lodash/isArrayLike';
+import keys from 'lodash/keys';
+
+var iteratorSymbol = typeof Symbol === 'function' && Symbol.iterator;
+
+export default function iterator(coll) {
+    var i = -1;
+    var len;
+    if (isArrayLike(coll)) {
+        len = coll.length;
+        return function next() {
+            i++;
+            return i < len ? {value: coll[i], key: i} : null;
+        };
+    }
+
+    if (iteratorSymbol && coll[iteratorSymbol]) {
+        var iterate = coll[iteratorSymbol]();
+
+        return function next() {
+            var item = iterate.next();
+            if (item.done)
+                return null;
+            i++;
```

Comment:
I really don't like tracking this index. People should pass `.entries` if they want access to the `key`. `i` may not correspond to the real index. Just leave `key` undefined